File: java/amp-commanding/src/main/java/amp/commanding/DefaultCommandBus.java
package amp.commanding;

import java.util.List;

import cmf.bus.IEnvelopeBus;
import amp.messaging.IMessageProcessor;
import amp.messaging.MessageException;

/**
 * Created with IntelliJ IDEA.
 * User: jar349
 * Date: 5/1/13
 * Time: 4:22 PM
 * To change this template use File | Settings | File Templates.
 */
public class DefaultCommandBus implements ICommandBus {

    private ICommandSender _sender;
    private ICommandReceiver _receiver;


    public DefaultCommandBus(IEnvelopeBus envelopeBus, 
    		List<IMessageProcessor> inboundProcessors,
    		List<IMessageProcessor> outboundProcessors) {
        _sender = new DefaultCommandSender(envelopeBus, outboundProcessors);
        _receiver = new DefaultCommandReceiver(envelopeBus, inboundProcessors);
    }


    @Override
    public void onCommandReceived(ICommandHandler handler) throws MessageException, IllegalArgumentException {
        _receiver.onCommandReceived(handler);
    }

    @Override
    public void send(Object command) throws MessageException {
        _sender.send(command);
    }

    @Override
    public void dispose() {
        _receiver.dispose();
    }
}

File: dotnet/source/amp.commanding/DefaultCommandBus.cs
using System;
using System.Collections.Generic;
using amp.messaging;
using cmf.bus;

namespace amp.commanding
{
    public class DefaultCommandBus : ICommandBus
    {
        protected readonly ICommandSender _commandSender;
        protected readonly ICommandReceiver _commandReceiver;


        public DefaultCommandBus(IEnvelopeBus envelopeBus
            , List<IMessageProcessor> inboundChain
            , List<IMessageProcessor> outboundChain)
        {
            _commandReceiver = new DefaultCommandReceiver(envelopeBus, inboundChain);
            _commandSender = new DefaultCommandSender(envelopeBus, outboundChain);
        }


        public void Send(object command)
        {
            _commandSender.Send(command);
        }

        public void Dispose()
        {
            try
            {
                _commandSender.Dispose();
            }
            catch
            {
            }

            try
            {
                _commandReceiver.Dispose();
            }
            catch
            {
            }
        }

        public void ReceiveCommand(ICommandHandler handler)
        {
            _commandReceiver.ReceiveCommand(handler);
        }

        public void ReceiveCommand<TCommand>(Action<TCommand, IDictionary<string, string>> handler) where TCommand : class
        {
            _commandReceiver.ReceiveCommand(handler);
        }
    }
}
